Add OIDC login

jniebuhr opened the pull request:
This PR adds OIDC capability to AKHQ. 
Username and roles will be taken from the OpenID claims and translated using the existing logic.
When activating OIDC login, the login page will show a redirect button instead of the login form.

jniebuhr commented:
Probably solves #142

tchiotludo commented:
Wohh !! 
Thanks for this one :+1: really appreciate ! 

Just some thinking : 
The most important : This PR will break this [one](https://github.com/tchiotludo/akhq/pull/298). 
I will change the whole ui with a new react ui. 
The main idea for now is to release a new version with both ui (old and new), this must be done this week I think (in order to gather bug).

The next version (after the one with both) will remove the old ui in favor in the react one. 
In this release, I will remove the session in favor of a JWT cookie (that will be compatible with your PR).
I think you will need to wait a little bit for this release in order to be merged. 

Other concern about the PR : 
- I want to be able to have both LDAP / basic auth with Oauth, so we won't disabled the login form if there is an auth 
- I want to be able to support multiple Oauth provider, so maybe `Oidc` must be `List<Oidc>`

What do you think about that ?

jniebuhr commented:
When the React UI is merged, I guess we'll need some configuration endpoint that returns which providers are enabled then.
For your other two concerns, I'll adjust the PR.

jniebuhr commented:
![Screenshot from 2020-07-27 11-10-34](https://user-images.githubusercontent.com/2579336/88524896-3d926780-cffa-11ea-9813-46fa7ceac296.png)

@tchiotludo 
I've changed it so you can configure multiple oidc providers and it doesn't hide the normal login if you do.

jniebuhr commented:
I think the check is broken? It's been at the same step for quite some time.

jniebuhr commented:
@tchiotludo 
It looks like you merge some stuff in for the new UI. I've resolved the conflicts for this PR. 
Should I also add the OIDC logic to the React UI? If yes, do you want to pass the configuration options to the frontend with an API endpoint?

tchiotludo commented:
Yes the old ui will be [remove](https://github.com/tchiotludo/akhq/pull/348) on next release ! 

Here you can find the [endpoint](https://github.com/tchiotludo/akhq/blob/3cd3c81d7f527830dd87f57f683581d4b1558293/src/main/java/org/akhq/controllers/AkhqController.java#L66-L76) where you can add the OpenId Providers

jniebuhr commented:
@tchiotludo there you go, implemented in react now

JacobJohansen commented:
I think Micronauts may need to be bumped to 1.3.7 to resolve an issue with host resolver https://github.com/micronaut-projects/micronaut-core/issues/3416

specifically when for the use-case when the proxy handles SSL termination

JacobJohansen commented:
> I think Micronauts may need to be bumped to 1.3.7 to resolve an issue with host resolver [micronaut-projects/micronaut-core#3416](https://github.com/micronaut-projects/micronaut-core/issues/3416)
> 
> specifically when for the use-case when the proxy handles SSL termination



> I think Micronauts may need to be bumped to 1.3.7 to resolve an issue with host resolver [micronaut-projects/micronaut-core#3416](https://github.com/micronaut-projects/micronaut-core/issues/3416)
> 
> specifically when for the use-case when the proxy handles SSL termination

scratch that, you have to enable forwarding headers; the documentation made it seem as if Micronauts did that by default

tchiotludo commented:
I confirm it's enabled by default.
But there is case you need to fine tune it, for example behind a GCP LB : 
```
    micronaut:
      server:
        host-resolution:
          host-header: Host
          protocol-header: X-Forwarded-Proto
```
Because micronaut expect both `X-Forwarded-Proto` & `X-Forwarded-Host` and Google LB don't send the `X-Forwarded-Host`

JacobJohansen commented on src/main/java/org/akhq/configs/Oidc.java:
```diff
@@ -0,0 +1,26 @@
+
+import java.util.Map;
+
+@ConfigurationProperties("akhq.security.oidc")
```
should this be a configuration each?

tchiotludo replied:
yes I think it must be `@EachProperty`

jniebuhr replied:
Nope, it is not an array so it has to be `@ConfigurationProperties`

tchiotludo replied:
my fault ;) 
To be honest, I don't have time to look deeply ;)

JacobJohansen commented on src/main/java/org/akhq/modules/OidcUserDetailsMapper.java:
```diff
@@ -0,0 +1,115 @@
+    protected Map<String, Object> buildAttributes(String providerName, OpenIdTokenResponse tokenResponse, OpenIdClaims openIdClaims) {
+        Map<String, Object> attributes = super.buildAttributes(providerName, tokenResponse, openIdClaims);
+        userGroupUtils.getUserAttributes(getOidcRoles(providerName, openIdClaims)).forEach(attributes::put);
+        attributes.remove("roles");
```
should this be the roles field?

JacobJohansen replied:
I was using the groups claim from Azur AD and It output to many AD/Office 365 groups for the cookie to store in the browser. Dropping the roles field cleaned up that space.

jniebuhr replied:
I‘m intentionally dropping a field called „roles“ because that is where Micronaut inserts the resolved roles. It will not do that if there is an attribute called „roles“.

For the other problem, rather than just copying everything over from the provider JWT we should probably add a whitelist.

tchiotludo replied:
To be honest, it's this part that hurt me, storing the roles in the JWT cookie.

In a future release, I will need to use another logic than role for akhq, people want to have role per cluster, that lead to incompatibility with the micronaut role.

I would have prefer to have only username in the JWT token, and server will add the role on the login part.

This will allow to change the role on server side without any impact on the cookie part and we won't have anymore the cookie size issue.

@jniebuhr what do you think about that ?

jniebuhr replied:
If we just Store the username in the cookie there won’t be a way to get the roles for the user, other than caching it in-memory or doing the OIDC flow all over again. We could store all the unresolved roles that had matches in the cookie, that should be smaller.

tchiotludo replied:
In another project, I use this to handle custom mapping : 
```java
import io.micronaut.security.oauth2.endpoint.token.response.DefaultOpenIdUserDetailsMapper;

@Singleton
@Replaces(DefaultOpenIdUserDetailsMapper.class)
public class GlobalOpenIdUserDetailsMapper extends DefaultOpenIdUserDetailsMapper {

    public GlobalOpenIdUserDetailsMapper(OpenIdAdditionalClaimsConfiguration openIdAdditionalClaimsConfiguration) {
        super(openIdAdditionalClaimsConfiguration);
    }

    @Override
    @Nonnull
    public UserDetails createUserDetails(String providerName, OpenIdTokenResponse tokenResponse, OpenIdClaims openIdClaims) {
        // user is here : openIdClaims.getEmail();
        // mapping username using configuration files 
        //  return new UserDetails();
    }
]
```
It works very well and nothing except the user is store on the cookie. 
The oauth flow is only done 1 time also.

jniebuhr replied:
Do you have a username to roles mapping in the Config then?

tchiotludo replied:
Yes but in this project I have a database to fetch the roles.

That lead to a question, where is the role in this actual implementation ?

jniebuhr replied:
The roles get passed in the provider token as a claim, I then map them to AKHQ roles. Those get stored in the JWT and can be accessed on every request.

tchiotludo replied:
So you have some openid provider where you can customize the claim content ? (add specific fields for an application) ?  

Most of users will use a generic openId provider (Google, Microsoft, ...) and will not be able to customize the claims (IMO)

jniebuhr replied:
You can’t always customize which claims are there but you can usually select them well enough. If the provider has no understanding about roles you’ll need to use a default role anyway. In our case we have full control over claims, yes.

jniebuhr replied:
Maybe we can add a default role for OIDC authenticated users

JacobJohansen replied:
what I have done in spring before but I believe it would be equivalent to replace the bean UserDetails and override getRoles to map the group "role" form the IDP to the RBAC roles that have been defined in the configuration. It will help keep the JWT token smaller. https://github.com/micronaut-projects/micronaut-security/blob/39eef7d3d3cfe6a25ba957026d725a78b769cbb8/security/src/main/java/io/micronaut/security/authentication/UserDetails.java

JacobJohansen replied:
> Most of users will use a generic openId provider (Google, Microsoft, ...) and will not be able to customize the claims (IMO)

to support this use-case there will probably need to be username to group "role" mapping in the future and then the UserDetails mapper can output the roles that are referenced on the endpoint annotation

tchiotludo replied:
Agree with that, but I also think it's a must have for OpenId.
Most of users will use a public / not customizable providers so we need to allow user to have a default role but also have a special mapping for username > roles for admin for example :) 
I will hope to have a better mapping (group based) but I don't see a better way for now :cry:

jniebuhr commented:
@tchiotludo @JacobJohansen I have reworked the whole users and groups mapping for OIDC to work like the LDAP one. I've also added a default group to both so they can share their code. I've renamed the ldap.user and ldap.group properties to ldap.users and ldap.groups but kept the old ones around for people to migrate.

tchiotludo commented:
Starting to look at deeply, seems to be a really good stuff here ! 
But I'll try to make it work with google oauth, I don't find for now. 
The redirect to callback is ok, but the custom mapper is not called, so the user has no role and will be redirect to login page. 

Don't understand for now :thinking:

tchiotludo commented:
To make it detect Mapper by micronaut, I must changed to this  : https://github.com/tchiotludo/akhq/commit/29f40564bc25685cf7d50904409c3ef6f32f6504

Maybe Google provider have not the same behaviour than yours ? 

I also changed the `GroupMapping` & `UserMapping` to a `List` since it don't work (think must be `@EachProperties` to work with a map (but I don't want to use Map in configuration files, since micronaut change the case of the key, it lead to a lot of bug like that https://github.com/tchiotludo/akhq/issues/336#issuecomment-663367630)

tchiotludo commented:
In a meantime, I've merged this one : 
https://github.com/tchiotludo/akhq/pull/380

Seems there is changed on the OpenId on the 2.0.0.

Also maybe, we should disabled session login to JWT everywhere ! AKHQ will work behind a load balancer without sticky session !
My only concern is `secret` (provide a default one or not). 

Maybe we must start to enable the JWT first to avoid a biggest PR ?

jniebuhr commented:
I'll check the whole implementation based on micronaut 2 and do a test with Google. I'll update the PR afterwards

jniebuhr commented:
> To make it detect Mapper by micronaut, I must changed to this : [29f4056](https://github.com/tchiotludo/akhq/commit/29f40564bc25685cf7d50904409c3ef6f32f6504)
> 
> Maybe Google provider have not the same behaviour than yours ?
> 
> I also changed the `GroupMapping` & `UserMapping` to a `List` since it don't work (think must be `@EachProperties` to work with a map (but I don't want to use Map in configuration files, since micronaut change the case of the key, it lead to a lot of bug like that [#336 (comment)](https://github.com/tchiotludo/akhq/issues/336#issuecomment-663367630))

If we change the Map to a List, it's gonna be a different syntax to the LDAP one

jniebuhr commented:
@tchiotludo I have changed the config to always use JWT now. I've tested the code with the pasted config, it works just fine using basic auth and oidc at the same time. It'll generate JWT tokens from the user details and store them in a cookie. A user upgrading to this will need to set a secret. For security reasons, we shouldn't provide a default one. If you want I can extract the JWT part from this PR, but it isn't a lot to make it work. Mostly adding the dependency and some config.

```
akhq:
  connections:
    plain:
      properties:
        bootstrap.servers: "kafka:9092"
      schema-registry:
        url: "http://schema-registry:8085"
      connect:
      - url: "http://connect:8083"
  security:
    default-group: no-roles
    basic-auth:
      demo:
        password: 2a97516c354b68848cdbd8f54a226a0a55b21ed138e207ad6c5cbb9c00aa5aea # demo
        groups:
          - admin
    oidc:
      enabled: true
      providers:
        google:
          label: "Login with Google"
          username-field: email
          groups-field: roles
          default-group: no-roles
          users:
            some-admin:
              username: <your gmail address here>
              groups:
                - admin

micronaut:
  server:
    port: 8080
  security:
    enabled: true
    oauth2:
      enabled: true
      clients:
        google:
          client-id: <client-id>
          client-secret: <client-secret>
          openid:
            issuer: "https://accounts.google.com"
    token:
      jwt:
        signatures:
          secret:
            generator:
              secret: thisisarandom256bitsecretyoullneedtochange

```

tchiotludo commented on src/main/resources/application.yml:
```diff
@@ -20,13 +20,18 @@ micronaut:
         mapping: "/swagger/**"
   security:
     enabled: true
```
Maybe we need to disabled security by default, since there is not `secret` defined and add some doc on readme about that (enabled security & add a secret)

jniebuhr replied:
Yes, that's a good idea i guess

tchiotludo commented on src/main/java/org/akhq/configs/Ldap.java:
```diff
@@ -0,0 +1,56 @@
+    /**
+     * @deprecated Naming isn't consistent with the other mappings
+     */
+    @Deprecated(forRemoval = true)
```
Go BC for a good time ! remove it ;) as `user`

jniebuhr replied:
Alright, if we go breaking that's no problem :+1:

tchiotludo commented on src/main/java/org/akhq/configs/Ldap.java:
```diff
@@ -0,0 +1,56 @@
+@Data
+public class Ldap {
+    private String defaultGroup;
+    private Map<String, GroupMapping> groups = new HashMap<>();
```
There is a lot of issue using a Map<String, T>, micronaut change the case of key and user complain about that.
I think we must go with a List<T> here  as `group`

jniebuhr replied:
Yeah, I only added it this way to stay consistent with the other entries. If we change it here, I'll change it in all occurences (oidc, ldap, basic auth, general groups)

jniebuhr replied:
There is a weird bug with `@EachProperty` in list mode. If you use it, it will not fill List<String> fields below it. I had to remove any `@EachProperty` and use a new `SecurityProperties` class for it.

tchiotludo reviewed (changes requested):
Added some minor change ! 
I would prefer to go breaking compatibility for the whole next release since we change the session to cookie, there is breaking change. (also remove the old ui) 
So I think we must go to a full breaking change !

tchiotludo commented:
I just try, it work perfectly ! :heart: 
The release must contains a big warning for all the breaking change :disappointed_relieved: but was necessary I think ! 
No more issue on this. 

Really thanks for the hard work and let people try it on dev branch before the next release ! 

:+1:

tchiotludo commented:
seeing your email, maybe you can contribute here also : https://github.com/tchiotludo/akhq/issues/161 :smile:
